src/retractions.jl

```diff
@@ -274,6 +274,9 @@ retract!(M::AbstractManifold, q, p, X, ::ExponentialRetraction) = exp!(M, q, p,
 function retract!(M::AbstractManifold, q, p, X, t::Real, method::AbstractRetractionMethod)
     return retract!(M, q, p, t * X, method)
 end
-function retract!(M::AbstractManifold, q, p, X, method::AbstractRetractionMethod)
-    return error(manifold_function_not_implemented_message(M, retract!, q, p, method))
-end
+
+#
+# dispatch to lower level
```

Comment:
Nonmutating variants should have a similar dispatch.

BTW, I think the nonmutating-through-mutating thing should happen on the lower level, it should be easier to overwrite at that point.

Reply:
Hm, I am not yet sure. I was thinking it should by default happen on the higher level, because it is less to implement. We would have at least 15 or so more function definitions (over all functions we switch to this scheme).

Why is it a better idea to do that on the lower level? I am not yet convinced.

Reply:
I think it's more suitable on the lower level because in case we need to override the default non-mutating implementation we don't have to add new manifold-specific methods to higher-level functions, which I think is a large part of the goal here -- implementation for a specific manifold should happen on the lower level. Especially for basis-related functions adding a manifold-specific specialization on the higher level would lead us back to the problems we currently have. Just look at methods of `get_vector` for `Circle` manifold.

Reply:
ah thats a good argument. Because my idea was exactly that, just „squeeze in“ a dispatch to the lower level, but you are right, that leads to the problems we want to resolve with this PR. I will add those, next time I work on this PR :)